Packet++/header/LdapLayer.h

```diff
@@ -569,7 +570,210 @@ namespace pcpp
+			: LdapLayer(std::move(asn1Record), data, dataLen, prevLayer, packet) {}
+
 		std::string getExtendedInfoString() const override;
+	private:
+		static constexpr int versionIndex = 0;
+		static constexpr int nameIndex = 1;
+		static constexpr int credentialIndex = 2;
+
+		static constexpr int saslMechanismIndex = 0;
+		static constexpr int saslCredentialsIndex = 1;
+	};
+
+	/**
+	 * @class LdapBindResponseLayer
+	 * Represents LDAP bind response operation
+	 */
+	class LdapBindResponseLayer : public LdapResponseLayer
+	{
+	public:
+		/**
+		 * A constructor to create a new LDAP bind response message
+		 * @param[in] messageId The LDAP message ID
+		 * @param[in] resultCode The LDAP result code
+		 * @param[in] matchedDN The distinguished name (DN) to set on the message. If not applicable
+		 * pass an empty string
+		 * @param[in] diagnosticMessage The additional information to set on the message. If not applicable
+		 * pass an empty string
+		 * @param[in] referral A list of URIs to re-try the operation somewhere else. This is an optional
+		 * parameter. If not provided then referral won't be added to the message
+		 * @param[in] serverSaslCredentials Encoded server SASL credentials for use in subsequent processing
+		 * @param[in] controls A vector of LDAP controls. This is an optional parameter, if not provided the message
+		 * will be created without LDAP controls
+		 */
+		LdapBindResponseLayer(uint16_t messageId, const LdapResultCode& resultCode, const std::string& matchedDN,
+			const std::string& diagnosticMessage, const std::vector<std::string>& referral = std::vector<std::string>(),
+			const std::vector<uint8_t>& serverSaslCredentials = std::vector<uint8_t>(),
+			const std::vector<LdapControl>& controls = std::vector<LdapControl>());
+
+		/**
+		 * @return Encoded server SASL credentials for use in subsequent processing
+		 */
+		std::vector<uint8_t> getServerSaslCredentials();
+	protected:
+		friend LdapLayer* LdapLayer::parseLdapMessage(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet);
+
+		static constexpr int serverSaslCredentialsTagType = 7;
+
+		LdapBindResponseLayer(std::unique_ptr<Asn1Record> asn1Record, uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet)
+			: LdapResponseLayer(std::move(asn1Record), data, dataLen, prevLayer, packet) {}
+	};
+
+	/**
+	 * @class LdapUnbindRequestLayer
+	 * Represents LDAP unbind operation
+	 */
+	class LdapUnbindRequestLayer : public LdapLayer
+	{
+	public:
+		// cppcheck-suppress noExplicitConstructor
+		/**
+		 * A constructor to create a new LDAP unbind message
+		 * @param[in] messageId The LDAP message ID
+		 * @param[in] controls A vector of LDAP controls. This is an optional parameter, if not provided the message
+		 * will be created without LDAP controls
+		 */
+		LdapUnbindRequestLayer(uint16_t messageId, const std::vector<LdapControl>& controls = std::vector<LdapControl>());
+
+		Asn1ConstructedRecord* getLdapOperationAsn1Record() const = delete;
```

Comment:
Yes I know it's a bit weird... LDAP unbind message is special because its operation record is empty. For example, here is a bind request operation message (ASN.1 representation):

![image](https://github.com/seladb/PcapPlusPlus/assets/9059541/59fb3d46-b313-4be4-bed4-bd2407578037)

And here is an unbind request operation:

![image](https://github.com/seladb/PcapPlusPlus/assets/9059541/7c4a831c-da62-4997-a2ea-d20982892591)

The operation message is everything under "Application". As you can see, the operation has sub-records in all other LDAP messages, which is what `getLdapOperationAsn1Record()` returns. Unbind's operation has no sub-records, hence calling `getLdapOperationAsn1Record()` will result in an exception. That's why I think it's better to delete it.

Another option could be to subclass `LdapMessage` and add `getLdapOperationAsn1Record()` as part of that subclass, then change the other message classes (except unbind) to inherit from that class. But I think it makes things even more complex and not worth the complexity. Please let me know what you think

Reply:
I see. Sure, if its guaranteed to throw an exception anyway, I suppose it is fine to delete it in the concrete class, if its a one off occurrence.

Reply:
Added a comment in this commit: https://github.com/seladb/PcapPlusPlus/pull/1448/commits/dba0d05220b4833d84314ffd0b0737d47f7f1cd7